### PR TITLE
[next-pwa]: add type declaration

### PR DESCRIPTION
**Pull request by NivaldoFarias**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by NivaldoFarias** [comment]
### Issues

There were a few errors and declaration conflicts caused by typescript native libraries, the main being the conflict between libs `dom` and `webworkers`, which for this package would be needed, but causes multiple conflict declaration errors. 

The error arose from needing a `PopStateEvent` declaration, which is only included inside the `dom` lib, while the remaining declarations where included inside the `webworker` lib. Removing `webworker` in favour of `dom` excludes multiple declarations specific to WebWorkers API, thus a no-go for this package. By looking through the [latest official libraries](https://github.com/microsoft/TypeScript/tree/main/lib) its noticeable how many declarations may cause conflicts _(dupes and overrides)_ when importing both libraries.

The [current workaround](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63391/files#diff-f953282badb249078078f78638e67692401a0f811c8195f7da8e21f369043079) was to declare globals to 'virtually' impose a declaration to the types with no matching declaration from the native libraries. Since this is more of a hack than a fix, I will try and update the package when *(if)* TypeScript's native libraries ecossystem is upgraded. 

###### See: [Discussion about DOM and WebWorker libs conflicts](https://github.com/microsoft/TypeScript/issues/20595)

**Review by martin-badin — CHANGES_REQUESTED** [review]

File: types/next-pwa/.eslintrc.json
Comment (NivaldoFarias) [reply types/next-pwa/.eslintrc.json:10]: I can't see to make these rules work inside my IDE environment, is it a common issue for the editor not to find the declaration for these rules? The screenshots below show the behaviour I have always encountered. `npm install` has been run on both the root repo and the type package repo.  

Index.d.ts eslint rules errors
next-pwa-tests.ts eslint rules errors

Comment (NivaldoFarias) [reply types/next-pwa/.eslintrc.json:10]: I removed the eslint config file and re-ran all tests and linting scripts, all seems to work still! Should I remove the `.eslintrc.json` file from the repo? The rules not working issue is probably related to a local environment bug, so removing it wouldn't affect the types package

Comment (martin-badin) [reply types/next-pwa/.eslintrc.json:10]: @NivaldoFarias Yes, please. You can remove the file

**Review by martin-badin — CHANGES_REQUESTED** [review]

File: types/next-pwa/index.d.ts
Comment (martin-badin) [inline types/next-pwa/index.d.ts:7]: Is this reference required?

Comment (NivaldoFarias) [reply types/next-pwa/index.d.ts]: Apparently not!! Guess I unknowingly used it as a workaround while trying to fix the mismatch to _DOM_ declarations issue [mentioned above](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63391#issuecomment-1328129756).

Comment (NivaldoFarias) [reply types/next-pwa/index.d.ts]: Apparently it does!! jadshgdasj minutes after I sent the previous reply, the CI failed. 

Refer to [these error logs](https://dev.azure.com/definitelytyped/DefinitelyTyped/_build/results?buildId=144987&view=logs&j=8635df7d-ecde-52a7-c7a8-84b998e35690&t=c264d665-0b79-59b7-7e04-9965cbe74860&l=129). Is it a common issue? If so, is there a default solution or should I re-introduce the reference to `@types/react`?

Comment (martin-badin) [reply types/next-pwa/index.d.ts:7]: Please. Add the React typings into package.json.

Comment (NivaldoFarias) [reply types/next-pwa/index.d.ts:7]: Awaiting [PR merge](https://github.com/microsoft/DefinitelyTyped-tools/pull/571) on `allowedPackageJsonDependencies.txt` in order to push the changes to the package _(remove `@types/react` triple-dash reference in favour of adding to package.json as a dependency)_

**Review by martin-badin — CHANGES_REQUESTED** [review]

**Review by martin-badin — APPROVED** [review]
looks good

File: types/next-pwa/index.d.ts
Comment (NivaldoFarias) [reply types/next-pwa/index.d.ts:132]: [Still broken](https://dev.azure.com/definitelytyped/DefinitelyTyped/_build/results?buildId=145119&view=logs&j=8635df7d-ecde-52a7-c7a8-84b998e35690&t=c264d665-0b79-59b7-7e04-9965cbe74860&l=92). How should I tackle this?

Comment (martin-badin) [reply types/next-pwa/index.d.ts:132]: maybe you can try this one
```
@default ["/!noprecache/**\/*"]
```

Comment (NivaldoFarias) [reply types/next-pwa/index.d.ts]: Tried it, but the `\` is picked up by the parser, probably because it is read as a string instead of a regex.

I chose to [remove the default value description](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63391/commits/4630df86b2a656024fcc0d51ce9162ea4fc8aacf) and instead include a reference to the original package's docs about the property.

Comment (NivaldoFarias) [reply types/next-pwa/index.d.ts]: I read through a few issues, mainly [tsdoc issue#166](https://github.com/microsoft/tsdoc/issues/166) and [jsdoc escape char mechanism](https://github.com/gajus/eslint-plugin-jsdoc/issues/710), and it seems that there is no solution, as of yet. It May be a change in the near future, so I'll be sure to keep an eye on it.

**Review by martin-badin — APPROVED** [review]

**Comment by NivaldoFarias** [comment]
Wasn't aware that the bot removes the `other approved` label when the forked repo is updated, I'll wait for a DT maintainer review and approval before updating the fork next time.

**Review by martin-badin — APPROVED** [review]

**Comment by martin-badin** [comment]
@NivaldoFarias Better way is to create a separate branch.

File: types/next-pwa/tsconfig.json
Comment (peterblazejewicz) [inline types/next-pwa/tsconfig.json:19]:
```suggestion
        "forceConsistentCasingInFileNames": true
```
https://github.com/DefinitelyTyped/DefinitelyTyped#esmoduleinteropallowsyntheticdefaultimports

Comment (NivaldoFarias) [reply types/next-pwa/tsconfig.json:19]: Disabling the `esModuleInterop` flag causes the following issues (refer to print).
 
image

Is there a known workaround to this? As of yet, I am only aware of resolving this issue by enabling the flag. Would love to know if there is a solution!

Comment (peterblazejewicz) [reply types/next-pwa/tsconfig.json:19]: there is no 'workaround' here required, JS module is CJS:
> refer to print

Please copy test you're pointing to next time, this is just an image. Do you get this error when testing? This module is a CJS. I'll try locally

Comment (peterblazejewicz) [reply types/next-pwa/tsconfig.json:19]: so the errors is due to `next` shipping ESM type of declaration files only, so when referenced here, spills the errors.
One should describe why changing default config (which is why 'check config' label was added) in first place, when starting PR.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!

File: types/next-pwa/tsconfig.json
Comment (NivaldoFarias) [reply types/next-pwa/tsconfig.json:19]: > One should describe why changing default config (which is why 'check config' label was added) in first place, when starting PR.

My bad! Before opening the PR the `tsconfig.json` had comments right above the `esModuleInterop` describing why it was enabled, which is not the default config. Forgot to add a comment to the **PR** after removing the `tsconfig.json` comment to comply to DT's package rules.

**Comment by NivaldoFarias** [comment]
@typescript-bot Ready to merge

**Pull request merged**
